blockdag/dag.go

```diff
@@ -1630,30 +1630,83 @@ func (dag *BlockDAG) blueBlocksBetween(lowHash, highHash *daghash.Hash, maxEntri
 	// first iterate on the selected parent chain of the highNode
 	// until we find a new highNode
 	// where highNode.blueScore-lowNode.blueScore+1 <= maxEntries
-
+	//
+	// We also keep track of highNode's selected parent chain child
+	// in case we need to calculate highNode's anticone (see below)
+	var highNodeSelectedParentChainChild *blockNode
 	for highNode.blueScore-lowNode.blueScore+1 > maxEntries {
+		highNodeSelectedParentChainChild = highNode
 		highNode = highNode.selectedParent
 	}
 
-	// Populate and return the found nodes.
-	nodes := make([]*blockNode, 0, highNode.blueScore-lowNode.blueScore+1)
-	nodes = append(nodes, highNode)
-	current := highNode
-	for current.blueScore > lowNode.blueScore {
-		for _, blue := range current.blues {
-			nodes = append(nodes, blue)
+	// Collect every node in highNode's past (including itself) but
+	// NOT in the lowNode's past (excluding itself) into an UpHeap
+	// (a heap sorted by blueScore from lowest to greatest).
+	visited := newSet()
+	highNodeAntiFuture := newUpHeap()
+	queue := newDownHeap()
+	queue.Push(highNode)
+	for queue.Len() > 0 {
+		current := queue.pop()
+		if visited.contains(current) {
+			continue
+		}
+		visited.add(current)
+		isCurrentAncestorOfLowNode, err := dag.isAncestorOf(current, lowNode)
+		if err != nil {
+			return nil, err
+		}
+		if !isCurrentAncestorOfLowNode {
+			highNodeAntiFuture.Push(current)
+			for _, parent := range current.parents {
+				queue.Push(parent)
+			}
 		}
-		current = current.selectedParent
 	}
-	if current != lowNode {
-		return nil, errors.Errorf("the low hash is not found in the " +
-			"selected parent chain of the high hash")
+
+	// If we collected less than maxEntries nodes in the above
+	// traversal then we also add highNode's anticone
+	if highNodeAntiFuture.Len() < int(maxEntries) {
+		// If we initially (from the function args) received
+		// highNode.blueScore-lowNode.blueScore+1 < maxEntries
+		// then we don't immediately know the selected parent
+		// chain child of highNode. To find it, we iterate
+		// over highNode's children until we stumble upon a
+		// child whose selectedParent is highNode.
+		if highNodeSelectedParentChainChild == nil {
+			for _, child := range highNode.children {
+				if child.selectedParent == highNode {
```

Comment:
There might be a few of those, and they are not necessarily the selected-parent-chain child.
You can use the selected-parent-chain-set to check for this.

However, I think (Please verify I am correct here) you don't actually need the glorified selected-parent-chain-child, but indeed, any node whose selectedParent is highNode will do. In that case - update comments and names to reflect this